### PR TITLE
DOC: Rearranged parts of the Indexing docs to consolidate content

### LINKED ISSUE #14038: DOC: Merge two indexing documentation pages
xref:https://github.com/numpy/numpy/issues/12195
Currently we have:

    https://docs.scipy.org/doc/numpy-1.15.0/user/basics.indexing.html
    https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html

I added a link between each other on pr https://github.com/numpy/numpy/pull/14025 but the info still remains a bit haywire and lot of the info could be merged.
Making it a issue so I could get any suggestions on how they could be improved. I will start working on this

EDIT: latest versions:
- https://numpy.org/devdocs/user/basics.indexing.html
- https://numpy.org/devdocs/reference/arrays.indexing.html

### PR DESCRIPTION
**Pull request by Mukulikaa**
Changes in [Indexing in NumPy Fundamentals](https://numpy.org/devdocs/user/basics.indexing.html):
1. Changed title from `Indexing` to `Indexing basics`.
2. Removed all major content from sections: Assignment vs referencing, Single element indexing, Other indexing options, Index arrays, Indexing Multi-dimensional arrays, Boolean or “mask” index arrays, Combining index arrays with slices, Structural indexing tools, and added them to the Indexing doc in the reference guide.
3. Kept very short overviews of the indexing options and cross-linked them to the Indexing doc in the reference guide.

Changes in [Indexing in API Reference](https://numpy.org/devdocs/reference/arrays.indexing.html):
1. Added information and examples from the basics doc (this required some subsections to be rearranged).
2. Moved Detailed notes section to the end of the doc to maintain continuity.
3. Added link to Indexing routines page

Closes #14038

Side note: I realize it's a huge PR and will potentially be difficult to review. Please let me know if there are better ways to go about these kinds of PRs that involve re-writes and changes to multiple files 🙂.

### CODE REVIEW

**Comment by Mukulikaa** [comment]
The CI build is failing because of two reasons:
1. I forgot to define a variable.
2. It is failing the example of an error. I am not sure if we can skip certain code blocks during testing. 

          >>> x = np.array([[1, 2], [3, 4], [5, 6]])
          >>> x[np.array([1, -1])]
          array([[3, 4],
                [5, 6]])
          >>> x[np.array([3, 4])]
          IndexError: index 3 is out of bounds for axis 0 with size 3`

**Comment by melissawm** [comment]
Hi @Mukulikaa - sorry for the delay!

A couple of high-level comments:

- I think the general idea of moving content from the user doc to the reference doc makes the [user doc](doc/source/user/basics.indexing.rst ) a little bit out of place as a "basics". In fact, reading it I kind of felt like it was way more advanced than the reference page, just because there are no examples and it is very abstract. 
- The original [Indexing basics](https://numpy.org/devdocs/user/basics.indexing.html#basics-indexing) page kind of reads like a verbose how-to; I think it's useful for beginners to have that kind of entry-level content about indexing. Many people are coming from other programming languages and get tripped up on slicing, advanced indexing etc. 
- The original [Indexing reference](https://numpy.org/devdocs/reference/arrays.indexing.html#arrays-indexing) is to me very much an [Explanation](https://diataxis.fr/explanation/) and, to me, should live in the Fundamentals section.
- There are some indexing examples in Quickstart and the Absolute Beginners documents, but they are also mixed in with a lot of other stuff.

So, here's my suggestion (but others may very well disagree!):
- Keep your edits to the `doc/source/reference/arrays.indexing.rst` document but move it to the fundamentals section;
- Maybe merge the basics (`doc/user/basics.indexing.rst`) document into the `arrays.indexing.rst` document
- Create an indexing how-to, with clear and concise examples of indexing operations.

Of course the how-to would come in another PR, but to me it makes more sense to organize things that way. 

Thoughts?

cc @mattip @rossbar

**Comment by Mukulikaa** [comment]
Thanks, @melissawm! I agree with all of your suggestions. I moved all the information to the reference guide doc because it was suggested [here](https://github.com/numpy/numpy/issues/14038#issuecomment-531498988) and I assumed users would prefer that. But I agree if we were to strictly follow Diataxis then we should move the doc to the Fundamentals. 

I am also in the process of making a list of use-cases for the how-to 🙂.

**Comment by melissawm** [comment]
So, right now here's what we have:
- In the root of NumPy Fundamentals we have [Indexing basics](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/user/basics.indexing.html)
- From Indexing basics, we can reach [Indexing](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.indexing.html#arrays-indexing) which is actually like a stub with other references for more complete documents.
- Now, in the reference docs, inside [The N-dimensional array we also have a subheader "Indexing arrays"](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.ndarray.html#indexing-arrays), which again just points to [Indexing](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.indexing.html#arrays-indexing)

I'm wondering if it would just be better to remove the [Indexing arrays](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.ndarray.html#indexing-arrays) section in the "N-dimensional array" page, and put that sentence directly into [Indexing](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.indexing.html).

Another option would be to lose the [Indexing](https://20930-908607-gh.circle-artifacts.com/0/doc/build/html/reference/arrays.indexing.html#arrays-indexing) page altogether; I'm struggling to understand if this is needed after the current changes.

Oh and another point: the "Indexing basics" page is not basic at all :) I'd change its title to Indexing or "Indexing on ndarrays", something like this.

**Comment by Mukulikaa** [comment]
> Another option would be to lose the Indexing page altogether; I'm struggling to understand if this is needed after the current changes.

I was also thinking about this because I agree that this doc is a stub. However, I felt that having that doc clearly visible in the sidebar makes it more accessible to someone who directly refers to the reference doc instead of the user guide (experienced users maybe?). So I was thinking if we could replace the stub with [Indexing routines](https://numpy.org/devdocs/reference/routines.indexing.html) which would link to the doc in the user guide. It will look something like the [ufuncs reference doc](https://21037-908607-gh.circle-artifacts.com/0/doc/build/html/reference/ufuncs.html) I restructured recently. 

> I'd change its title to Indexing or "Indexing on ndarrays", something like this.

I agree, we can change it to "Indexing on ndarrays" to avoid any clash.

> I'm wondering if it would just be better to remove the Indexing arrays section in the "N-dimensional array" page, and put that sentence directly into Indexing.

I don't have any particular opinion on this so I'm happy to amend it the way everyone thinks is best!

**Comment by Mukulikaa** [comment]
The doctests are still failing on code from an old commit. Maybe CircleCI needs a re-run?

**Comment by Mukulikaa** [comment]
I am not sure if the CI failure is unrelated.

**Comment by mattip** [comment]
This is fixed on `main` but unfortunately circleCI does not "merge-from-main" before building a PR. So you will have to get those changes into this PR manually. I would recommend rebasing the PR on top of `main`, you could also merge `main` into the PR. A rebase would look like
```
git checkout main
git pull
git checkout indexing-docs-merge
git rebase
# if there are conflicts, you need to do some cycles of 
# git status
# <see what has conflicts>
# fix the file: look for sections with "<<<<", "=====", ">>>>>" and resolve the conflicts
# git add <file>
# git rebase --continue
git push -f Mukulikaa
```

**Comment by Mukulikaa** [comment]
[Here](https://21533-908607-gh.circle-artifacts.com/0/doc/build/html/user/basics.indexing.html) is the rendered page.

File: doc/source/user/basics.indexing.rst [hunk 1]
```diff
@@ -39,8 +43,7 @@ and accepts negative indices for indexing from the end of the array. ::
     >>> x[-2]
     8
 
-Unlike lists and tuples, NumPy arrays support multidimensional indexing
-for multidimensional arrays. That means that it is not necessary to
+It is not necessary to
 separate each dimension's index into its own set of square brackets. ::
 
     >>> x.shape = (2,5) # now x is 2-dimensional
```
Comment (mattip) [inline doc/source/user/basics.indexing.rst:49]: Maybe scope creep, but could you reformat these examples with spaces after the commas?

Comment (mattip) [reply doc/source/user/basics.indexing.rst:49]: Also below (I cannot comment directly there), the sentence 
> It must be noted that the returned array is not a copy of the original,
but points to the same values in memory as does the original array.

should have the glossary term "view" worked in (as a link) since that is the official term

**Comment by mattip** [comment]
Hmm, I started going through the content, but on second thought maybe that should be left for a follow-on PR. This does such a great job moving the pieces around it would be a shame to delay merging it on smaller details.

**Comment by Mukulikaa** [comment]
If it is easier to leave comments on this PR for the changes required, I am happy to continue the work on this PR itself. As I am actively working on this it wouldn't cause much delay I think. 🙂

File: doc/source/user/basics.indexing.rst [hunk 2]
```diff
@@ -59,313 +62,627 @@ That is, each index specified selects the array corresponding to the
...
+          [29, 30]])
+
+
+.. _arrays.indexing.fields:
+
+Field access
+-------------
+
+.. seealso:: :ref:`arrays.dtypes`, :ref:`arrays.scalars`,
+            :ref:`structured_arrays`
```
Comment (rossbar) [inline doc/source/user/basics.indexing.rst:631]: The first two links are ported over from the original document, but TBH I think the link you've added is by far the most relevant.

Again, not to overload this PR too much, but I'd be +1 for removing the first two links and leaving only the structured array link here.

File: doc/source/user/basics.indexing.rst [hunk 3]
```diff
@@ -405,17 +722,19 @@ example is often surprising to people: ::
  array([ 0, 11, 20, 31, 40])
 
 Where people expect that the 1st location will be incremented by 3.
-In fact, it will only be incremented by 1. The reason is because
+In fact, it will only be incremented by 1. The reason is that
 a new array is extracted from the original (as a temporary) containing
 the values at 1, 1, 3, 1, then the value 1 is added to the temporary,
 and then the temporary is assigned back to the original array. Thus
-the value of the array at x[1]+1 is assigned to x[1] three times,
+the value of the array at x[1] + 1 is assigned to x[1] three times,
```
Comment (rossbar) [inline doc/source/user/basics.indexing.rst:729]:
```suggestion
the value of the array at ``x[1] + 1`` is assigned to ``x[1]`` three times,
```

File: doc/source/user/basics.indexing.rst [hunk 4]
```diff
@@ -424,7 +743,7 @@ supplies to the index a tuple, the tuple will be interpreted
 as a list of indices. For example (using the previous definition
 for the array z): ::
 
- >>> indices = (1,1,1,1)
+ >>> indices = (1, 1, 1, 1)
```
Comment (rossbar) [inline doc/source/user/basics.indexing.rst:746]: This references an array `z` that no longer exists (or at least is very far away from the current example).
```suggestion
 >>> z = np.arange(81).reshape(3, 3, 3, 3)
 >>> indices = (1, 1, 1, 1)
```
Note also that the text above the example should be modified to reflect this change.

**Review by rossbar — CHANGES_REQUESTED** [review]
This is really impressive work @Mukulikaa , distilling the disparate info from these (heavily overlapping) documents is no small feat! I reviewed this with my main goal being to check that all of the relevant information from the two original documents ([ref guide article](https://numpy.org/devdocs/reference/arrays.indexing.html#) & [user guide article](https://numpy.org/devdocs/user/basics.indexing.html#). For the most part I think this is satisfied - there were a few places where I thought some text/examples from one or the other document that had been dropped should be reincluded.

The only main thing that jumped out to me was the expansion of indexing types to include "single element indexing" as a top-level category. As mentioned in the specific comments, this should be folded under the "basic" indexing category. This was the only necessary change though IMO, everything else is either opinions or nits. Thanks for the excellent work here!

**Comment by Mukulikaa** [comment]
Thank you for your comments @rossbar! I think I omitted some examples to reduce the length of the document but I agree that they should be included if they make the concepts clearer.

**Comment by Mukulikaa** [comment]
Hi @rossbar, I've tried to incorporate the changes you suggested. Let me know if there are any other suggestions!

File: doc/source/user/basics.indexing.rst [hunk 5]
```diff
@@ -59,313 +68,704 @@ That is, each index specified selects the array corresponding to the
...
+Combining advanced and basic indexing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When there is at least one slice (``:``), ellipsis (``...``) or :const:`newaxis`
+in the index (or the array has more dimensions than there are advanced indices),
+then the behaviour can be more complicated. It is like concatenating the
+indexing result for each advanced index element.
+
+In the simplest case, there is only a *single* advanced index combined with
+a slice. For example:
```
Comment (rossbar) [inline doc/source/user/basics.indexing.rst:616]:
```suggestion
a slice. For example::
```

File: doc/source/user/basics.indexing.rst [hunk 6]
```diff
@@ -59,313 +68,704 @@ That is, each index specified selects the array corresponding to the
...
+``x[ind_1, boolean_array, ind_2]`` is equivalent to
+``x[(ind_1,) + boolean_array.nonzero() + (ind_2,)]``.
+
+If there is only one Boolean array and no integer indexing array present,
+this is straightforward. Care must only be taken to make sure that the
+boolean index has *exactly* as many dimensions as it is supposed to work
+with.
+
+In general, when the boolean array has fewer dimensions than the array being
+indexed, this is equivalent to ``x[b, …]``, which means x is indexed by b
```
Comment (rossbar) [inline doc/source/user/basics.indexing.rst:521]:
```suggestion
indexed, this is equivalent to ``x[b, ...]``, which means x is indexed by b
```
A nuanced nit, but it looks like the unicode character for an elipsis was used here instead of the 3 `.` characters

**Review by rossbar — APPROVED** [review]
Thanks for the iteration @Mukulikaa , you've addressed all my original comments very well - just a few more things I noticed in the updated text but otherwise this looks good to me. Excellent work!

File: doc/source/user/basics.indexing.rst [hunk 7]
```diff
@@ -1,35 +1,45 @@
...
 
+Note that in Python, ``x[(exp1, exp2, ..., expN)]`` is equivalent to
+``x[exp1, exp2, ..., expN]``; the latter is just syntactic sugar
+for the former.
+
+
+.. _basic-slicing-and-indexing:
+
+Basic slicing and indexing
+--------------------------
```
[the remark this answers is not in the document]
Comment (Mukulikaa) [reply doc/source/user/basics.indexing.rst:35]: It hasn't been used anywhere else so I'll go ahead and change it.

File: doc/source/user/basics.indexing.rst [hunk 8]
```diff
@@ -1,35 +1,45 @@
...
-when assigning to an array. See the section at the end for
+when assigning to an array. See :ref:`assigning-values-to-indexed-arrays` for
 specific examples and explanations on how assignments work.
 
+Note that in Python, ``x[(exp1, exp2, ..., expN)]`` is equivalent to
+``x[exp1, exp2, ..., expN]``; the latter is just syntactic sugar
+for the former.
+
+
+.. _basic-slicing-and-indexing:
```
Comment (Mukulikaa) [inline doc/source/user/basics.indexing.rst:32]:
```suggestion
.. _basic-indexing:
```

File: doc/source/user/basics.indexing.rst [hunk 9]
```diff
@@ -1,35 +1,45 @@
...
+
+Basic slicing and indexing
```
Comment (Mukulikaa) [inline doc/source/user/basics.indexing.rst:34]:
```suggestion
Basic indexing
```

File: doc/source/user/basics.indexing.rst [hunk 10]
```diff
@@ -29,9 +29,9 @@ Note that in Python, ``x[(exp1, exp2, ..., expN)]`` is equivalent to
 for the former.
 
 
-.. _basic-slicing-and-indexing:
+.. _basic-indexing:
```
Comment (mattip) [inline doc/source/user/basics.indexing.rst:32]: This link may be in use by others outside NumPy. Could we leave both forms with a comment "legacy link name"?

**Comment by mattip** [comment]
The latest commit which changed a link name made me start thinking about a side-effect of this refactoring: it might break existing links. I took a look at the [nitpicky](https://github.com/numpy/numpy/issues/13114) CI run [results](https://app.circleci.com/pipelines/github/numpy/numpy/9679/workflows/7419b74d-606b-4dca-814c-e99301ff8b0d/jobs/21778?invite=true#step-107-5238) for internal consistenc. It seems the WARNINGS: with the word "index" in them (like the See Also section of [expand_dims](https://numpy.org/devdocs/reference/generated/numpy.expand_dims.html) and missing `numpy.lib.index_tricks` in the second paragraph of [np.choose](https://numpy.org/devdocs/reference/generated/numpy.choose.html)) existed before this PR and no new ones appeared. But there may be content out there that does rely on the older links. Should we make an effort to preserve them where possible?

File: doc/source/user/basics.indexing.rst (hunk 10 above)
Comment (Mukulikaa) [reply doc/source/user/basics.indexing.rst:32]: I added this link in this PR itself in a previous commit because I needed it for the stub Indexing doc that was left behind in the reference guide. Since then I removed the stub completely but forgot to remove this link... So should I just remove this link altogether?

Comment (mattip) [reply doc/source/user/basics.indexing.rst:32]: Ahh, I see. If it is new and no-one needs it then it can be removed.

**Comment by mattip** [comment]
Thanks @Mukulikaa. Let's put this in, we can make more changes if it turns out we broke links by rearranging things.
